### ISSUE TITLE
socket.io fails with TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be one of type string, Buffer, or URL

### ISSUE THREAD

**Issue by styfle** [zeit/ncc#322]
It looks like ncc doesn't work with `socket.io`.

This is probably related to: #148 and #200

Original report from @travismiller in [this comment](https://github.com/zeit/now-builders/issues/284#issuecomment-474638887)

## Steps to reproduce

```sh
git clone https://github.com/styfle/ncc-bug-socketio
cd ncc-bug-socketio
yarn install
ncc build index.js
node dist/index.js
```

## Error

```
internal/fs/utils.js:453
    throw err;
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be one of type string, Buffer, or URL. Received type function
    at Object.openSync (fs.js:430:3)
    at readFileSync (fs.js:343:35)
    at Server.module.exports.117.Server.serveClient (/Users/styfle/Desktop/ncc-bug-socketio/dist/index.js:1638:20)
    at new Server (/Users/styfle/Desktop/ncc-bug-socketio/dist/index.js:1575:8)
    at Server (/Users/styfle/Desktop/ncc-bug-socketio/dist/index.js:1566:41)
    at Object.318 (/Users/styfle/Desktop/ncc-bug-socketio/dist/index.js:4568:1)
    at __webpack_require__ (/Users/styfle/Desktop/ncc-bug-socketio/dist/index.js:22:30)
    at startup (/Users/styfle/Desktop/ncc-bug-socketio/dist/index.js:36:19)
    at /Users/styfle/Desktop/ncc-bug-socketio/dist/index.js:40:18
    at Object.<anonymous> (/Users/styfle/Desktop/ncc-bug-socketio/dist/index.js:43:10)
```

**Comment by guybedford** [comment]
Ok so the failure here is the following code in socket-io:

```js
Server.prototype.serveClient = function(v){
  if (!arguments.length) return this._serveClient;
  this._serveClient = v;
  var resolvePath = function(file){
    var filepath = path.resolve(__dirname, './../../', file);
    if (exists(filepath)) {
      return filepath;
    }
    return require.resolve(file);
  };
  if (v && !clientSource) {
    clientSource = read(resolvePath( 'socket.io-client/dist/socket.io.js'), 'utf-8');
    try {
      clientSourceMap = read(resolvePath( 'socket.io-client/dist/socket.io.js.map'), 'utf-8');
    } catch(err) {
      debug('could not load sourcemap file');
    }
  }
  return this;
};
```

where the line that is causing the error is `clientSource = read(resolvePath( 'socket.io-client/dist/socket.io.js'), 'utf-8');`.

So here's where the analysis is:
* It knows that `read` is fs.readFileSync because that was set at the top of the source fine - `var read = require('fs').readFileSync;`
* It does not know that the string 'socket.io-client/dist/socket.io.js' is a path string that should be analyzed.
* It does not know to try and use the local `resolvePath` implementation, as we aren't a fully interpreted compiler.

If we wanted to improve the interpreted analysis to catch this kind of asset case, that might be possible, I'm not sure.

Basically we would need to extend the analysis to handle:
1. FS functions as asset analysis triggers. This would incur some perf cost analyzing call use of FS functions.
2. During analysis, to be able to lazily determine binding values. This would be a refactoring of the current logic which precomputes its known bindings rather than trying to look them up as needed.
3. Supporting functions and statements in the static analysis. Again this has some perf cost risks, as well as potential security risks.

It's not a small undertaking, but it might be possible.

**Comment by guybedford** [comment]
The alternative here might be to just make this another "special case" and specifically do a transformation just for this code that directly works out where the assets are and emits them properly.

Perhaps that is the best approach really.

One thing with these special cases is to ensure that we keep the tests updated, because small internal code changes could break our handling of them. So we should take that into account in the testing methodology.

**Comment by zandaqo** [comment]
I'm having the same problem with `passport` or one of the related strategy modules, i.e. it appears when I include OAuth, I have not had time to debug yet, but here is the log:
```
internal/validators.js:125
    throw new ERR_INVALID_ARG_TYPE(name, 'string', value);
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type undefined
    at validateString (internal/validators.js:125:11)
    at Object.dirname (path.js:646:5)
    at Function.module.exports.pkginfo.find (D:\dist\index.js:51904:14)
    at Function.module.exports.pkginfo.read (D:\dist\index.js:51930:17)
    at module.exports.module.exports (D:\\dist\index.js:51881:21)
    at Object.<anonymous> (D:\dist\index.js:77569:25)
    at __webpack_require__ (D:\dist\index.js:23:30)
    at Object.<anonymous> (D:\dist\index.js:45450:27)
    at __webpack_require__ (D:\dist\index.js:23:30)
    at Object.<anonymous> (D:\dist\index.js:45046:18)

Process finished with exit code 1
```

**Comment by guybedford** [comment]
@zandaqo thanks for reporting. Despite the message being the same, the case is likely slightly different - if you're able to post a new issue with some simple replication details that would be amazing to look into it further. Otherwise if it is as simple as `npm install passport` and building it directly then I can try that out too.

**Comment by zandaqo** [comment]
@guybedford A bit more digging points towards `passport-microsoft` and its dependency `[pkginfo]`(https://www.npmjs.com/package/pkginfo). From the looks of it, `pkginfo` wants to read `package.info` file of its dependent and is unable to find it. If that's the case, I thinks it's not so much of a ncc problem as just a poor choice of those using `pkginfo`, so I wouldn't rush with opening a ticket here.

A way to replicate this shoud be to install:

```npm i passport passport-microsoft```

request them both:

```
const passport = require('passport');
const MicrosoftStrategy = require('passport-microsoft').Strategy;
```
 
and compile with `ncc index.js -o dist`.

**Comment by guybedford** [comment]
@zandaqo thanks for the update, I've created https://github.com/zeit/ncc/issues/348 to track this specific issue.

**Comment by guybedford** [comment]
PR for this created in https://github.com/zeit/webpack-asset-relocator-loader/pull/17.
